routing_strategy.go

```diff
@@ -11,39 +11,42 @@ type RoutingStrategy interface {
 	SelectRoute(ctx context.Context,
 		req Request,
 		routes map[string]Component,
-	) (route Component, fallbacks []Component, err error)
+	) (route Component, fallbacks []Component, labels Labels, err error)
 }
 
 type baseRoutingStrategy struct {
 	RoutingStrategy
 	BaseFiberType
 }
 
+type routesOrderResponse struct {
+	Components []Component
+	Labels     Labels
+	Err        error
+}
+
 func (s *baseRoutingStrategy) getRoutesOrder(
 	ctx context.Context,
 	req Request,
 	routes map[string]Component,
-) (<-chan []Component, <-chan error) {
-	out := make(chan []Component)
-	errCh := make(chan error, 1)
+) <-chan routesOrderResponse {
+	out := make(chan routesOrderResponse)
 
 	go func() {
-		route, fallbacks, err := s.SelectRoute(ctx, req, routes)
-
-		if err != nil {
-			errCh <- err
-		} else {
-			// Append routes
-			routes := fallbacks
-			if route != nil {
-				routes = append([]Component{route}, routes...)
-			}
-			out <- routes
+		route, fallbacks, labels, err := s.SelectRoute(ctx, req, routes)
+
+		// Combine preferred route with the fallbacks
+		routes := fallbacks
```

Comment:
probably not related to this PR, but I think this part is abit fragile or with implicit assumption that fallback is an array returned.

Intuitively to me, scenarios are 
1. No primary nor fallback
2. Primary with no fallback
3. Primary with fallbacks.

For scenario 2, it make sense to me that its nil-able, since there are no fallbacks as there is only one route. This is expecting fallbacks to be returned as an empty slice when there is no fallback nor error.

It make more sense to create an empty slice, append route and fallbacks if they are not nil instead of using `routes := fallback` as returning nil when there are no fallback seems valid to me.

Hope i didnt understood anything wrong 😅

Reply:
Your understanding of the scenarios is correct. In addition, there can also be:
4. No primary route, with fallback (which we use in [Turing](https://github.com/caraml-dev/turing/blob/main/engines/router/missionctl/fiberapi/routing_strategy.go#L124) - the fallback will always be set if there is a default route.). Although the handling of the primary vs fallbacks is not very different upstream, in the Fiber component.

Thanks for raising the concern about the `nil` slice. I tried to test my understanding and it's actually not a problem. This article summarizes the behavior: https://freshman.tech/snippets/go/nil-vs-empty-slices/#whats-the-difference

The in-built functions all work the same, even for a `nil` slice. The part that differs apparently is if your JSON encode it.